modules/runner/src/main/java/org/apache/ignite/internal/configuration/ServiceLoaderModulesProvider.java

```diff
@@ -0,0 +1,33 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.configuration;
+
+import java.util.ServiceLoader;
+import java.util.stream.Stream;
+
+/**
+ * Provides {@link ConfigurationModule}s using {@link ServiceLoader} mechanism.
+ * Uses {@link ConfigurationModule} interface to query the ServiceLoader.
+ *
+ * @see ConfigurationModule
+ */
+public class ServiceLoaderModulesProvider {
+    public Stream<ConfigurationModule> modules() {
```

Comment:
I would suggest to use a simple List as the return type, it makes a more clear interface

Reply:
Why is `List` clearer than `Stream`?

`Stream` has an advantage (subjectively) that it does not force you to think about the return type as about something that is indexable, for example. It is just a sequence (probably lazy) of elements.

Reply:
I think it over complicates things in this case: most user would simply expect a collection of modules, not a lazily populated stream that you have to manually collect into a list or whatever.

Reply:
`Stream` also might need to be closed, so the clients would need to think: should we close it or not? So `List` is less troubling, switched to it.